packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,326 @@
+# Nimble Table Columns API
+
+## Problem Statement
+
+The `nimble-table` requires users to be able to configure which columns to display for a table. Additionally, a particular display column may require access to multiple fields in the data, and allow a user to define the field to sort by. We need to provide a means for a client to provide their own column visualization. Finally, columns must also provide the user to either specify the header text, or custom visuals to use for the header.
+
+### Out of scope of this HLD
+
+Programmatic API for state that could be considered column-centric: width, sort direction, grouped, etc. These concerns should be covered in separate designs covering those topics specifically, allowing for discussion on both the interactive side and the API design on an individual basis.
+
+## Links To Relevant Work Items and Reference Material
+
+[Table Spec](./README.md)
+
+[Table Declarative Columns API prototype:](https://github.com/ni/nimble/blob/325983040e886e52a100664d8fb1129dee767c2f/packages/nimble-components/src/table/tests/table.stories.ts#L23) ([Storybook](https://60e89457a987cf003efc0a5b-twewjutggo.chromatic.com/iframe.html?args=&id=table--table-story&viewMode=story))
+
+## Implementation / Design
```

Comment:
It would be good to explicitly mention some of the decisions around framework integration:
1. columns can be provided only via FAST-based custom elements, not Angular or Blazor elements
2. we expect Nimble will provide most standard column types (text, numeric, link, etc) and we'll include framework-specific integrations (e.g. Angular directives) for them

And I'm curious if anyone has opinions about whether `nimble-components` should re-export relevant FAST types so that someone can build a custom column outside the Nimble monorepo without depending directly on FAST. I'd lean towards yes.

Reply:
> And I'm curious if anyone has opinions about whether nimble-components should re-export relevant FAST types so that someone can build a custom column outside the Nimble monorepo without depending directly on FAST.

My initial reaction is I don't think nimble should be responsible for exposing the api surface for implementing new custom elements. If you want to make a new fast component use the FAST apis, i.e. we didn't create some massive abstraction on top of fast for component dev either, nor do I want to.

Maybe you are suggesting we just expose TableColumn and the types for the `html` helper and the `css` helper? If that's it then maybe we could just make them static on TableColumn

```ts
TableColumn {
    static readonly html = html;
    static readonly css = css;
}
```

But I dunno there are all the other [directives](https://www.fast.design/docs/fast-element/using-directives) commonly used like `repeat`, `when`, etc. yea I dunno if it's worth defining a subset like that.

I lean towards external custom column configuration plugins depending on fast directly. The issue could be if there is misalignment on fast versions used.. If there are issues with that then maybe the right thing to do would be to make fast a peerDependency in nimble (a peer so there is globally one version if it's not possible to have multiple versions in a page). I think fast is designed with the intent that multiple versions can live on a page as long as they don't register with the same name. This case might be an exception though as we need to share implementation-specific objects between systems (ie the output of the `html` and `css` tagged template helpers).

Reply:
I've added a section below briefly covering framework integration expectations, which covers @jattasNI's first two points. I propose we table the other discussion regarding how we expect clients to deal with the dependencies required for implementing a `TableColumn`. Ultimately that is tangential to the proposed APIs.

Reply:
@rajsite I had only been thinking about Nimble exporting the FAST types directly on Nimble's interfaces (`ViewTemplate` and `ElementStyles` were as far as I thought through). But that's a great point that we'd also need `html`, `css`, and all the directives you need to construct a template. I agree that's way too much for Nimble to re-export.

I'm happy to defer this discussion indefinitely / until we have a client who wants to write a column outside of this repo.